### ISSUE TITLE
[OPTIMIZATION] Methods that return 3 or more possible runtime-types optimize very poorly

### ISSUE THREAD

**Issue by techsy730** [vigna/fastutil#183]
Something discovered when optimizing Guava is that Hotspot (OpenJDK's virtual machine) optimizes megamorphic variables/returns (those that could be 3 or more different runtime types) _far_ worse then it does mono or dimorphic ones (1 or 2 different implementations). Sometimes by factors in the double digits. (https://github.com/google/guava/issues/1268)

As such, we should avoid static utility methods that can return potentially 3 or more different implementations.

This may (or may not) be better in newer JVMs, but in Java 8, the one we are targeting, it is absolutely an issue.

**Comment by vigna** [comment]
Mmmmhhh... I think there are a lot of those. Like, the various .of() methods which optimize for an empty set, a singleton, etc.

**Comment by techsy730** [comment]
There are two offenders that immediately come to mind
* The `of` static factories from the root interfaces (`List.of(...)`, etc)
* The wrappers of Spliterator in `Spliterators` converting to/from object based and primitive iterators

**Comment by techsy730** [comment]
The `of` methods can be resolved by having the empty case return a "well known" empty instance of the kind the varargs one returns. Such as `ImmutableList.EMPTY` instead of `Lists.EMPTY` for `List.of()`.

The `Iterators` case will probably require me to flatten the class hierarchy, pulling the `instanceof` into `IteratorWrapper` instead of having a subclass of it. (would want to measure this impact)

**Comment by vigna** [comment]
Yeah, I thought about that. We can have an immutable empty list exposed directly by ImmutableList. I'm writing that.

**Comment by techsy730** [comment]
I've already done that in https://github.com/vigna/fastutil/pull/180

**Comment by vigna** [comment]
Great! :) I'm still finding unused classes like

PrimitiveSpliteratorWrapperWithComparator

in ByteSpliterators.

**Comment by techsy730** [comment]
Oops, I missed those. Forgot to check the shorter types' warnings. I'll fix those

**Comment by techsy730** [comment]
List.of and Set.of are the most important.

Iterators.asXIterator and its cousin Spliterators.asXSpliterator 
However, the performance loss only happens if three or more returns actually happen in a program run, not just possible at all. And I can't imagine that many programs will trigger all three paths. Still should be looked at but less important.

**Comment by techsy730** [comment]
I should try to measure which has the worst impact performance wise:
1. The current status quo (megamorphic return)
2. Moving the `instanceof` check into the nextInt() method
3. Not trying to handle a JDK PrimitiveIterator at all and just have that go through that box/unbox process as well even though we could avoid it

**Comment by vigna** [comment]
I think we can consider this done, right?

**Comment by techsy730** [comment]
Not quite. There is still the case.

> - The wrappers of Iterator/Spliterator in Iterators/Spliterators converting to primitive ones from object based ones (e.g. asIntIterator).

But that is a pretty niche usage and I don't think it is worth blocking 8.5.0 over.

**Comment by vigna** [comment]
OK, I'll leave this open then.
